code/datums/components/archaeology.dm

```diff
@@ -0,0 +1,97 @@
+		If your turf has a unique post-dig sprite like basalt/asteroid, put that in postdig_icon as a text string AND set postdig_icon_change = TRUE.
+*/
+/datum/component/archaeology
+	dupe_type = COMPONENT_DUPE_UNIQUE
+	var/list/drops = list()
+	var/prob2drop = 0
+	var/mob/user
+	var/obj/item/W
+	var/dug = FALSE
+
+/datum/component/archaeology/Initialize(_prob2drop)
+	prob2drop = Clamp(_prob2drop, 0, 100)
+	if(isopenturf(parent))
+		var/turf/open/OT = parent
+		drops = OT.archdrops
+		RegisterSignal(COMSIG_PARENT_ATTACKBY,.proc/Dig)
+
+/datum/component/archaeology/Destroy()
+	user = null
+	W = null
+	return ..()
+
+/datum/component/archaeology/InheritComponent(datum/component/archaeology/A, i_am_original)
+	var/list/other_drops = A.drops
+	var/list/_drops = drops
+	for(var/I in other_drops)
+		_drops[I] += other_drops[I]
+
+/datum/component/archaeology/proc/Dig(mob/user, obj/item/W)
+	if(dug)
+		to_chat(user, "<span class='notice'> Looks like someone has dug here already.</span>")
+		return
+	else
+		var/digging_speed
+		if (istype(W, /obj/item/shovel))
+			var/obj/item/shovel/S = W
+			digging_speed = S.digspeed
+		else if (istype(W, /obj/item/pickaxe))
+			var/obj/item/pickaxe/P = W
+			digging_speed = P.digspeed
+
+		if (digging_speed && isturf(user.loc))
+			to_chat(user, "<span class='notice'>You start digging...</span>")
+			playsound(parent, 'sound/effects/shovel_dig.ogg', 50, 1)
+
+			if(do_after(user, digging_speed, target = parent))
+				to_chat(user, "<span class='notice'>You dig a hole.</span>")
+				gets_dug()
+				dug = TRUE
+				return TRUE
+		return FALSE
+
+/datum/component/archaeology/proc/gets_dug()
+	if(dug)
+		return
+	else
+		if(isopenturf(parent))
+			var/turf/open/OT = parent
+			for(var/thing in drops)
+				var/maxtodrop = drops[thing]
+				for(var/i in 1 to maxtodrop)
+					if(prob(prob2drop)) // can't win them all!
+						new thing(OT)
+
+			if(OT.postdig_icon_change)
+				if(istype(OT, /turf/open/floor/plating/asteroid/) && !OT.postdig_icon)
+					var/turf/open/floor/plating/asteroid/AOT = parent
+					AOT.icon_plating = "[AOT.environment_type]_dug"
+					AOT.icon_state = "[AOT.environment_type]_dug"
+				else
+					if(isplatingturf(OT))
+						var/turf/open/floor/plating/POT = parent
+						POT.icon_plating = "[POT.postdig_icon]"
+					OT.icon_state = "[OT.postdig_icon]"
+
+			if(OT.slowdown) //Things like snow slow you down until you dig them.
+				OT.slowdown = 0
+			SSblackbox.add_details("pick_used_mining",W.type)
+	dug = TRUE
+	user = null
+	W = null
+
+/******************************************************
+***************** MISC ******************************/
+
+/datum/component/archaeology/basalt
+
+/datum/component/archaeology/basalt/gets_dug()
+	..()
+	if(istype(parent,/turf/open/floor/plating/asteroid/basalt))
+		var/turf/open/floor/plating/asteroid/basalt/BT
+		BT.set_light(0)
```

Comment:
Instead of adding a component subtype. On the basalt turf, you can put this in a proc like this

```
/turf/open/floor/plating/asteroid/basalt/ComponentActivated(datum/component/C)
    ..()
    if(istype(C, /datum/component/archaeology))
        set_light(0)
```

Reply:
Would this trigger even if the do_after fails since it's just listening for the signal though?